docs/design.md

```diff
@@ -0,0 +1,164 @@
+
+### External components
+
+- Object storage: Store logical backups. It must have Amazon S3 compatible APIs (e.g. Ceph RGW).
+
+### Tools
+
+- `kubectl-myso`: CLI to manipulate MySQL cluster. It provides functionalities such as:
+  - Change master manually.
+  - Port-forward to MySQL servers.
+  - Execute SQL like `mysql -u -p` without a credential file on a local environment.
+  - Fetch a credential file to a local environment.
+
+### Diagram
+
+![design_architecture](http://www.plantuml.com/plantuml/png/ZPC_Rzim4CLtVeg3koI3WLhQYg88ean6qhWLWg2mF53InLPDaIf9fJMAVFTIF5b_R0lfOe3lFTtlZZxU6CkrhKl1U6QLXYU3cgrZrAz5XebySLB9ZKMz0MwWD6IS98pWhsXymXmJJAeSuQlx-TZeisUbzwXdy9S1L3pRyot2tSfRTrW9TIXHMh81rkN1g8qt7mxnumm-_40Xr6zdlh6tGUt5CVFj7XvydrS4kXwJufYIiUKMNseHuuZrI-GBmj5X6XWCW74pzc8A6Bm33FuF69u2WobW-0umd0RsPQAb3qLABbNQ5KYTMrSRy5vQIA6sYSphiZhhE-UMfzmWh6EhjAj8q4GcqZMOgRBYjaW59l8n_OcPXJRM15on-sAxixUYHaS-teBhnMebLe9BfEP80AzGZnrsngkYMK7KuseqHoJgRfTdxU4g-dxjigj2xYbqwuUDintcjK3AZhMpJmQNJlc1C0f6sbyiXTf35w0RR5uWpETVkiS6dy3z-Rrk9lHs7YT76Xs-TyRTzu7cbPry-injjYFZ_CjpVp3dfR8qOEyCuv_dy-PPNj222rh8scB-72qPhQIZ73V8KPugeNZEkejpReDmpaghFmTcnaOmPOJi1flr5ly0)
+
+The operator has the responsibility to create master-slave configuration of MySQL clusters.
+
+When the `MySQLCluster` is created, the operator starts deploying a new MySQL cluster as follows.
+In this section, the name of `MySQLCluster` is assumed to be `mysql`.
+
+1. The operator creates `StatefulSet` which has `N`(`N`=3 or 5) `Pod`s and its headless `Service`.
+1. The operator sets `mysql-0` as master and the other `Pod`s as slave.
+   The index of the current master is managed under `MySQLCluster.status.currentMasterIndex`.
+1. The operator creates some k8s resources.
+  - `Service` for accessing master.
+  - `Service` for accessing slaves.
+  - `Secrets` to store credentials.
+  - `ConfigMap` to store cluster configuration.
+
+Behaviors
+---------
+
+MySO is implemented as a custom controller, so it includes custom resource definitions(CRD).
+
+### How to execute failover when the master fails
+
+When the master fails, the cluster is recovered in the following process:
+
+1. Stop the `IO_THREAD` of all slaves.
+2. Select and configure new master.
+3. Update `MySQLCluster.status.currentMasterIndex` with the new master name.
+4. Turn off read-only mode on the new master
+
+In the process, the operator configures the old master as slave if the server is ready.
+
+### How to execute failover when a slave fails
+
+When a slave fails once and it restarts afterwards, the operator configures it to follow the master.
+
+### How to execute switchover
+
+Users can execute master switchover by applying `SwitchoverJob` CR which contains the master index to be switched to.
+
+Note that while any `SwitchoverJob` is running, another `SwitchoverJob` can be created but the operator waits for the completion of running jobs.
```

Comment:
What happens If a user submits multiple SwitchoverJobs which have different masterIndex simultaneously?
(or multiple administrator user submit multiple jobs.)
This is a some kind of mistake, however, operator should reject or validate it.

Reply:
In internal discussion, two jobs will be executed in any serialized order, so administrator should submit the job carefully.